### PR TITLE
Added WindowClosingEventArgs.

### LINKED ISSUE #9524: API to distinguish between user closing a window vs programmatically
**Is your feature request related to a problem? Please describe.**

Currently there is no way in the `Window.Closing` event to determine whether a window was closed by code or by user interaction.

App policy for what needs to be done in these two cases can be very different, so it's important to be able to distinguish these two cases. 

**Describe the solution you'd like**

Two potential API shapes are:

1. The `EventArgs` parameter to `OnClosing` could include an enum that indicates the cause of the close (user or code calling `Close()`)
2. Make `Window.Close(bool)` public, so a programmatic close could call `Close(ignoreCancel: true)`, skipping `OnClosing` entirely.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Adds `WindowClosingEventArgs` and uses it for `Window.Closing`. This API allows the event listener to detect:

- Why the window is closing
- Whether the close was done programmatically or via user interaction

Note that `WindowCloseReason.OSShutdown` is included but not currently supported by any platform backends. Not sure how we should handle this:

- Implement it as part of this PR?
- Remove the enum value for now?
- Add a note that not all platform backends may support this enum value and may send `ApplicationShutdown` instead?

## Fixed issues

Fixes #9524.

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027611-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Controls/WindowClosingEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        /// <summary>
+        /// The window is closing due to the application shutting down.
+        /// </summary>
+        ApplicationShutdown,
+
+        /// <summary>
+        /// The window is closing due to the operating system shutting down.
+        /// </summary>
+        OSShutdown,
```
Comment (maxkatz6) [inline src/Avalonia.Controls/WindowClosingEventArgs.cs:33]: Let's remove it for now

Comment (grokys) [reply src/Avalonia.Controls/WindowClosingEventArgs.cs:33]: My only reasoning for keeping it was: it's possible that someone has a check in their code such as `if (e.CloseReason == WindowCloseReason.ApplicationShutdown)` - if we then later add `OSShutdown` then that code is possibly no longer going to work correctly.

Having said that, I'm fine with removing it.

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027712-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**